docker_cli/image.py

```diff
@@ -0,0 +1,458 @@
+
+        inspect_result = run(
+            cmd,
+            capture_output=True,
+            text=True,
+            check=True
+        )
+
+        output_text = inspect_result.stdout
+        return output_text
+
+    def run(
+        self,
+        command: str,
+        *,
+        stdout: Optional[Union[io.TextIOBase, int]] = None,
+        stderr: Optional[Union[io.TextIOBase, int]] = None,
+        interactive: bool = False,
+        network: str = "host"
+    ):
+        """
+        Run the given command on a container.
+
+            .. note::
+        For stderr special values to pass into `stdout` and `stdin`, review
+        values passable into the same arguments in the :func:`~subprocess.run`
+        function:
+        https://docs.python.org/3/library/subprocess.html#subprocess.run
+
+            .. warning::
+        This method does not work correctly if the image built does not have
+        bash installed.
+
+        Parameters
+        ----------
+        cmd : str
+            The desired command, in linux command format.
+        stdout : io.TextIOBase or subprocess special value or None, optional
+            The location to send the process stdout output to.Defaults to None.
+        stderr : io.TextIOBase or subprocess special value or None, optional
+            The location to send the process stderr output to.
+            Defaults to None.
+        interactive : bool, optional
+            A boolean describing whether or not to run this command in
+            interactive mode or not. Defaults to True.
+        network : str, optional
+            The name of the network. Defaults to "host".
+
+        Returns
+        -------
+        str, optional
+            The output of the process, if `stdout` == PIPE.
+
+        Raises
+        -------
+        CalledProcessError
+            If the docker inspect command fails.
+        CommandNotFoundOnImageError:
+            When a command is attempted that is not recognized on the image.
+        """
+        cmd = ["docker", "run", f"--network={network}", "--rm"]
+        if interactive:
+            cmd += ["-i"]
+            if stdin.isatty():
+                cmd += ["--tty"]
+        cmd += [self._id, "bash"]
+        cmd += ["-ci"] if interactive else ["-c"]
+        cmd += split(f"'{command}'")
+
+        try:
+            result = run(
+                cmd,
+                text=True,
+                stdout=stdout,  # type: ignore
+                stderr=stderr,  # type: ignore
+                check=True
+            )
+            retval = result.stdout
+        except CalledProcessError as err:
+            if err.returncode == 127:
+                raise CommandNotFoundOnImageError(err, split(command)[0])
+            else:
+                raise err
+        return retval
+
+    def check_command_availability(self, commands: Sequence[str]) -> List[str]:
+        """
+        Determines which of the commands in a list are present on the image.
```

Comment:
I did this but decided to cast the output iterable to a list in my update in order to preserve how it interfaces with the other code.

I've realized just now that this method ends up being frustratingly slow in practice, even if it's not a bottleneck per se - it spins up several entire docker containers in order to do its task and that seems to take a pretty long time.

I wonders if we might benefit from a Container class that governs  an open container on a known Image and can be used to rapidly execute commands on it, or if that would overly complicate the interface for little benefit.

Reply:
I mentioned this on slack earlier today but I'll repeat the main thrust here for posterity.

The `Container` class idea is a really good one from a design perspective. It's an intuitive abstraction and fits well with the existing design.

However, I think this might be a case where [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) applies. At this point, I don't think it's likely that the overhead from spinning up the same container multiple times will be a significant bottleneck for us, and I'm hesitant to add another way of doing things without a stronger justification.

So I'd say we should keep this idea in our back pocket for now and revisit it in the future if we find that our need for something like this outweighs the cost of having to maintain it.